tardis/adapters/sites/satellite.py

```diff
@@ -0,0 +1,381 @@
+import asyncio
+import logging
+import ssl
+from contextlib import contextmanager
+from functools import partial
+from typing import Optional
+
+import aiohttp
+
+from tardis.exceptions.tardisexceptions import TardisResourceStatusUpdateFailed
+from tardis.interfaces.siteadapter import ResourceStatus, SiteAdapter
+from tardis.utilities.attributedict import AttributeDict
+from tardis.utilities.staticmapping import StaticMapping
+from tardis.utilities.asynccachemap import AsyncCacheMap
+
+logger = logging.getLogger("cobald.runtime.tardis.interfaces.site")
+
+
+class SatelliteClient:
+    """
+    Async helper for interacting with Satellite instance.
+    """
+
+    def __init__(
+        self,
+        host: str,
+        username: str,
+        secret: str,
+        ssl_cert: str,
+        machine_pool: list[str],
+        max_age: int,
+    ) -> None:
+
+        self._base_url = f"https://{host}/api/v2/hosts"
+        self.ssl_context = ssl.create_default_context(cafile=ssl_cert)
+        self.auth = aiohttp.BasicAuth(username, secret)
+
+        self.machine_pool = machine_pool
+
+        self.max_age = max_age * 60
+        self.cached_status_coroutines = {}
+
+    def _host_url(self, remote_resource_uuid: str = "") -> str:
+        if remote_resource_uuid == "":
+            return f"{self._base_url}/"
+        resource = remote_resource_uuid.strip("/")
+        return f"{self._base_url}/{resource}"
+
+    async def _request(
+        self,
+        session: aiohttp.ClientSession,
+        method: str,
+        url: str,
+        *,
+        expect_json: bool = True,
+        headers={
+            "Accept": "application/json",
+            "Foreman-Api-Version": "2",
+        },
+        **kwargs,
+    ):
+        async with session.request(
+            method,
+            url,
+            ssl=self.ssl_context,
+            headers=headers,
+            **kwargs,
+        ) as response:
+            response.raise_for_status()
+            if expect_json:
+                return await response.json()
+            return None
+
+    async def get_status(self, remote_resource_uuid: str, *, force: bool = False):
+        """
+        Return cached version of self._get_status().
+
+        :param remote_resource_uuid: Satellite identifier of the host.
+        :type remote_resource_uuid: str
```

Comment:
General comment:
If you use typehints, there is no need to specify the type again in the doc string.